File: javascript/ransom_note.js
function checkMagazine(magazine, note) {
    
    // loop thru magazine. insert each word in hashtable. if word already exist in hash just increment it.
    let hash = {}

    for (let word of magazine) {
        if (word in hash) {
            hash[word]++
        } else {
            hash[word] = 1
        }
    }
    // console.log(hash)

    // loop thru note. check if word not in hash or word is equal to 0. print No and return.
    for (let word of note) {
        if (!hash[word] || hash[word] === 0) {
            console.log("No")
            return
        } else {
            // else we want to decrement the word in hash
            hash[word]--
        }
    }
    
    // if we make it all the way thru, that means we can make the note out of magazine. print Yes.
    console.log("Yes")
}

checkMagazine(["give", "me", "me", "me", "give", "night"], ["give", "me", "me", "me", "give"])

File: python/ransom_note.py
def checkMagazine(magazine, note):
    # check if length of note is greater than length magazine
    if len(note) > len(magazine):
        print("No")
        return

    hash = {}
    # loop thru magazine and store it in hash.
    for word in magazine:
        if word in hash:
            hash[word] += 1 
        else:
            hash[word] = 1
    
    # loop thru the note and check if it's in the hash
    for word in note:
        if word not in hash or hash[word] == 0:
            print("No")
            return
        else:
            hash[word] -= 1
    
    print("Yes")

checkMagazine(["give", "me", "me", "me", "give", "night"], ["give", "me"])  
checkMagazine(["give", "me", "me", "me", "give", "night"], ["give", "me", "me", "me", "give"])        
checkMagazine(["give", "me", "me", "me", "give", "night"], ["give", "me", "me", "me", "give", "dog"])
checkMagazine(["give", "me", "me", "me", "give", "night"], ["Give", "me", "me", "me", "give", "nigth"])
checkMagazine(["give", "me", "me", "me", "give", "night"], ["give", "me", "me", "me", "give", "night", "night"])
checkMagazine(["give", "me", "me", "me", "give", "night"], ["give", "me", "me", "me", "me"])